File: java/src/main/java/org/libpetri/core/TokenInput.java
package org.libpetri.core;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Consumed input tokens bound to their source places.
 *
 * <p>Passed to {@link TransitionAction#execute} as the {@code in} parameter,
 * providing type-safe read access to tokens consumed from input places.
 *
 * <h3>Example</h3>
 * <pre>{@code
 * .action((in, out) -> {
 *     Request request = in.value(requestPlace);
 *     out.add(responsePlace, process(request));
 *     return CompletableFuture.completedFuture(out);
 * })
 * }</pre>
 *
 * @see TokenOutput for collecting output tokens
 */
public final class TokenInput {
    private final Map<Place<?>, List<Token<?>>> tokens = new HashMap<>();

    /**
     * Add a token (used by executor when firing transition).
     */
    public <T> TokenInput add(Place<T> place, Token<T> token) {
        tokens.computeIfAbsent(place, _ -> new ArrayList<>()).add(token);
        return this;
    }

    /**
     * Get all tokens for a place.
     */
    @SuppressWarnings("unchecked")
    public <T> List<Token<T>> getAll(Place<T> place) {
        var list = tokens.get(place);
        if (list == null) return List.of();
        return (List<Token<T>>) (List<?>) Collections.unmodifiableList(list);
    }

    /**
     * Get the first token for a place.
     * @throws IllegalArgumentException if no tokens for this place
     */
    @SuppressWarnings("unchecked")
    public <T> Token<T> get(Place<T> place) {
        var list = tokens.get(place);
        if (list == null || list.isEmpty()) {
            throw new IllegalArgumentException("No token for place: " + place.name());
        }
        return (Token<T>) list.getFirst();
    }

    /**
     * Get the first token's value for a place.
     * @throws IllegalArgumentException if no tokens for this place
     */
    public <T> T value(Place<T> place) {
        return get(place).value();
    }

    /**
     * Get all token values for a place.
     */
    public <T> List<T> values(Place<T> place) {
        return getAll(place).stream().map(Token::value).toList();
    }

    /**
     * Get token count for a place.
     */
    public int count(Place<?> place) {
        return getAll(place).size();
    }

    /**
     * Check if any tokens exist for a place.
     */
    public boolean has(Place<?> place) {
        return !getAll(place).isEmpty();
    }
}

File: typescript/src/core/token-input.ts
import type { Place } from './place.js';
import type { Token } from './token.js';

/**
 * Consumed input tokens bound to their source places.
 *
 * Passed to TransitionAction as the `input` parameter, providing
 * type-safe read access to tokens consumed from input places.
 */
export class TokenInput {
  private readonly tokens = new Map<string, Token<any>[]>();

  /** Add a token (used by executor when firing transition). */
  add<T>(place: Place<T>, token: Token<T>): this {
    const existing = this.tokens.get(place.name);
    if (existing) {
      existing.push(token);
    } else {
      this.tokens.set(place.name, [token]);
    }
    return this;
  }

  /** Get all tokens for a place. */
  getAll<T>(place: Place<T>): readonly Token<T>[] {
    return (this.tokens.get(place.name) ?? []) as Token<T>[];
  }

  /** Get the first token for a place. Throws if no tokens. */
  get<T>(place: Place<T>): Token<T> {
    const list = this.tokens.get(place.name);
    if (!list || list.length === 0) {
      throw new Error(`No token for place: ${place.name}`);
    }
    return list[0] as Token<T>;
  }

  /** Get the first token's value for a place. Throws if no tokens. */
  value<T>(place: Place<T>): T {
    return this.get(place).value;
  }

  /** Get all token values for a place. */
  values<T>(place: Place<T>): readonly T[] {
    return this.getAll(place).map(t => t.value);
  }

  /** Get token count for a place. */
  count(place: Place<any>): number {
    return this.getAll(place).length;
  }

  /** Check if any tokens exist for a place. */
  has(place: Place<any>): boolean {
    return this.count(place) > 0;
  }
}
